File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    List<String> path = new LinkedList<>();
    List<List<String>> ans = new ArrayList<>();

    public List<List<String>> partition(String s) {
        backtracking(s, 0);
        return ans;
    }

    private void backtracking(String s, int startIdx) {
        if (startIdx >= s.length()) {
            ans.add(new ArrayList<>(path));
            return;
        }

        for (int i = startIdx; i < s.length(); i++) {
            if (!isPalindrome(s, startIdx, i)) {
                continue;
            }

            path.add(s.substring(startIdx, i + 1));
            backtracking(s, i + 1);
            path.removeLast();
        }
    }

    private boolean isPalindrome(String s, int i, int j) {
        while (i < j) {
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        ans = []

        def isValid(s: str, i: int, j: int) -> bool:
            while i < j:
                if s[i] != s[j]:
                    return False
                i += 1
                j -= 1
            
            return True
        
        def backtracking(path: List[str], index: int) -> None:
            if index == len(s):
                ans.append(path[:])
                return
            
            for i in range(index, len(s)):
                word = s[index: i + 1]
                if isValid(s, index, i):
                    path.append(word)
                    backtracking(path, i + 1)
                    path.pop()
            
        backtracking([], 0)
        return ans